google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenCensusUtil.java

```diff
@@ -0,0 +1,174 @@
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.TraceId;
+import io.opencensus.trace.TraceOptions;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import io.opencensus.trace.samplers.Samplers;
+
+import java.util.Base64;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+final class OpenCensusUtil {
+  private static final Logger logger = Logger.getLogger(OpenCensusUtil.class.getName());
+
+  public static final String TAG_CONTEXT_KEY = "googclient_OpenCensusTagContextKey";
+  public static final String TRACE_CONTEXT_KEY = "googclient_OpenCensusTraceContextKey";
+
+  private static final Tagger tagger = Tags.getTagger();
+  private static final TagContextBinarySerializer serializer =
+      Tags.getTagPropagationComponent().getBinarySerializer();
+
+  private static final TraceOptions SAMPLED = TraceOptions.builder().setIsSampled(true).build();
+  private static final Tracer tracer = Tracing.getTracer();
+
+  // Used in Publisher.
+  // TODO(dpo): add configuration support to control adding these attributes.
+  static PubsubMessage putOpenCensusAttributes(PubsubMessage message) {
+    return PubsubMessage.newBuilder(message)
+        .putAttributes(
+            TRACE_CONTEXT_KEY,
+            encodeSpanContext(tracer.getCurrentSpan().getContext()))
+        .putAttributes(
+            TAG_CONTEXT_KEY,
+            encodeTagContext(tagger.getCurrentTagContext()))
+        .build();
+  }
+
+  // Used in Subscriber.
+  static MessageReceiver createOpenCensusMessageReceiver(MessageReceiver receiver) {
+    return new OpenCensusMessageReceiver(receiver);
+  }
+
+  private static String encodeSpanContext(SpanContext ctxt) {
+    TraceId traceId = ctxt.getTraceId();
+    SpanId spanId = ctxt.getSpanId();
+    TraceOptions traceOpts = ctxt.getTraceOptions();
+    return "traceid=" + traceId.toLowerBase16()
+        + "&spanid=" + spanId.toLowerBase16()
+        + "&traceopt=" + (traceOpts.isSampled() ? "t&" : "f&");
+  }
+
+  private static String encodeTagContext(TagContext tags) {
+    try {
+      byte[] encodedTags = serializer.toByteArray(tags);
+      return Base64.getEncoder().encodeToString(encodedTags);
+    }
+    catch (TagContextSerializationException exn) {
+      logger.log(Level.INFO, "OpenCensus: Tag Context Serialization Exception: " + exn);
+      return "";
+    }
+  }
+
+  private static Scope createScopedTagContext(String encodedTags) {
+    try {
+      TagContext tags = serializer.fromByteArray(Base64.getDecoder().decode(encodedTags));
+      return tagger.withTagContext(tags);
+    } catch (TagContextDeserializationException exn) {
+      logger.log(Level.INFO, "OpenCensus: Tag Context Deserialization Exception: " + exn);
+      return tagger.withTagContext(tagger.getCurrentTagContext());
+    }
+  }
+
+  @MustBeClosed
+  private static Scope createScopedSpan(String name) {
+    return tracer
+        .spanBuilderWithExplicitParent(name, tracer.getCurrentSpan())
```

Comment:
What would be the current span on the receiver side? Shouldn't it start a new root span?